### PR TITLE
Expand component search indexing fields

### PR DESCRIPTION
**Pull request by Mbeaulne**
## Description

Expands the component search index to include richer input/output details and a new `metadata` match field.

Previously, the `io` searchable field only contained input and output names. It now includes descriptions, types, and annotations for each input and output spec. A new `metadata` field has been added that indexes component-level metadata annotations (with a blocklist for noisy keys like `python_original_code`, editor state, and similar large/irrelevant blobs) as well as the source label and `published_by` value from the component reference.

The `MatchField` type and all related scoring, labeling, and UI display logic have been updated to include `metadata` alongside the existing fields. Annotation values longer than 500 characters are excluded from indexing to avoid polluting search with large blobs.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [x] Improvement
- [ ] Cleanup/Refactor
- [ ] Breaking change
- [ ] Documentation update

## Checklist

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Open the component dashboard and search for a term that appears in a component's metadata annotations (e.g. a framework name like `sklearn` or `lightgbm`).
2. Verify the result surfaces with `metadata` listed as a matched field.
3. Search for a publisher email address and confirm the matching component appears.
4. Search for a term that exists only in `python_original_code` or other excluded annotation keys and confirm it does **not** return results.
5. Search for an input/output description or type (e.g. `parquet`, `artifact`) and confirm results appear with `io` as the matched field.

## Additional Comments

The annotation exclusion list (`ANNOTATION_KEYS_EXCLUDED_FROM_SEARCH`) and the 500-character value length cap are the primary mechanisms for keeping the metadata index clean. These can be extended as new noisy annotation keys are identified.

### CODE REVIEW

**Comment by Mbeaulne** [comment]
* **#2462** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2462?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2461** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2461?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2459** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2459?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2458** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2457** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2456** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2456?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2454** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2453** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2453?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2452** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2452?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2451** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2451?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2450** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2450?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2449** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2449?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2448** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2447** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2447?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2444** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2433** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2433?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2432** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2432?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2431** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2430** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2429** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2428** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2427** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2427?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2426** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2426?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2425** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2425?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2424** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2424?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2423** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2423?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2423?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by camielvs** [comment]
## 🤖 Code review — Expand component search indexing fields

Reviewed as the base of the AI-search stack (#2423→#2433). Overall this is a clean, well-tested expansion: input/output **descriptions + types + annotations** now feed the `io` field, and a new `metadata` field indexes component-level annotations, the source label, and `published_by`. Excluded-key set and the 500-char per-value cap sensibly keep large blobs (`python_original_code`, editor positions) out. The `extractAnnotationsText` / `stringifySearchValue` helpers are tidy and the test coverage matches the new behavior well.

A few small things worth a look — none blocking:

- **`source.label` indexed into `metadata` makes generic tokens match broadly.** Because the source label is folded into the searchable `metadata` text, typing `user`, `standard`, or `published` now matches *every* component from that source (at weight 1). On short queries this can inject noise. The `ComponentSearchSource.id` doc comment already anticipates "future filter chips / URL state" — source feels more like a filter facet than a free-text token. Worth confirming this is the intended UX.

- **No aggregate cap on annotation text.** `extractAnnotationsText` caps each *value* at `MAX_ANNOTATION_TEXT_LENGTH` (500) but there's no bound on the *number* of annotations concatenated. A component with many sub-500-char annotations could produce a large searchable string. Bounded in practice for real components, but a total-length guard would make the index size predictable.

- **`published_by` is an email and is now searchable.** It's already surfaced in the UI (`ComponentHistoryTimeline`, `ComponentItem`), so this is consistent rather than a new disclosure — just flagging that searching by author email is now possible by design.

- **Question:** `python_dependencies` is in the excluded-keys set. Users sometimes search by library/dependency ("tensorflow", "lightgbm"). Implementation text (image + command) covers some of this, but was excluding dependencies a deliberate signal/size tradeoff?

Nice cleanup folding the duplicated name/type guards into `isNonEmptyString`.

File: src/services/componentSearchIndex.ts [hunk 1]
```diff
@@ -175,10 +245,12 @@ export function buildSearchIndex(sourced: SourcedReference[]): IndexEntry[] {
       searchable: {
         name: metadata.name.toLowerCase(),
         description: metadata.description.toLowerCase(),
-        io: [...metadata.inputNames, ...metadata.outputNames]
+        io: metadata.ioText.toLowerCase(),
+        implementation: extractImplementationText(reference),
+        metadata: [metadata.metadataText, source.label, reference.published_by]
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: **`source.label` indexed into `metadata` makes generic tokens match broadly.** Because the source label is folded into the searchable `metadata` text, typing `user`, `standard`, or `published` now matches *every* component from that source (at weight 1). On short queries this can inject noise. The `ComponentSearchSource.id` doc comment already anticipates "future filter chips / URL state" — source feels more like a filter facet than a free-text token. Worth confirming this is the intended UX.

_(Moved inline from my stack review on #2423 so it can be discussed in a thread.)_

File: src/services/componentSearchIndex.ts [hunk 2]
```diff
@@ -72,6 +77,59 @@ export function indexEntryToLexicalMatch(entry: IndexEntry): LexicalMatch {
...
+  annotations: Record<string, unknown> | undefined,
+): string {
+  if (!annotations) return "";
+
+  const parts: string[] = [];
+  for (const [key, value] of Object.entries(annotations)) {
+    if (ANNOTATION_KEYS_EXCLUDED_FROM_SEARCH.has(key)) continue;
+
+    const valueText = stringifySearchValue(value).trim();
+    if (!valueText || valueText.length > MAX_ANNOTATION_TEXT_LENGTH) continue;
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts:127]: **No aggregate cap on annotation text.** `extractAnnotationsText` caps each *value* at `MAX_ANNOTATION_TEXT_LENGTH` (500) but there's no bound on the *number* of annotations concatenated. A component with many sub-500-char annotations could produce a large searchable string. Bounded in practice for real components, but a total-length guard would make the index size predictable.

_(Moved inline from my stack review on #2423.)_

File: src/services/componentSearchIndex.ts (hunk 1 above)
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: **`published_by` is an email and is now searchable.** It's already surfaced in the UI (`ComponentHistoryTimeline`, `ComponentItem`), so this is consistent rather than a new disclosure — just flagging that searching by author email is now possible by design.

_(Moved inline from my stack review on #2423.)_

File: src/services/componentSearchIndex.ts [hunk 3]
```diff
@@ -72,6 +77,59 @@ export function indexEntryToLexicalMatch(entry: IndexEntry): LexicalMatch {
   };
 }
 
+const ANNOTATION_KEYS_EXCLUDED_FROM_SEARCH = new Set([
+  "editor.position",
+  "editor.collapsed",
+  "editor.flow-direction",
+  "flex-nodes",
+  "python_dependencies",
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: **Question:** `python_dependencies` is in the excluded-keys set. Users sometimes search by library/dependency ("tensorflow", "lightgbm"). Implementation text (image + command) covers some of this, but was excluding dependencies a deliberate signal/size tradeoff?

_(Moved inline from my stack review on #2423.)_

File: src/services/componentSearchIndex.ts (hunk 1 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — source labels are no longer folded into free-text metadata indexing. The source is still attached to each index entry for UI badges/future facets, but queries like "user", "standard", or "published" won't broadly match source groups anymore.

File: src/services/componentSearchIndex.ts (hunk 2 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts:127]: > 🤖 This is an AI-generated response.

Fixed — annotation indexing now has an aggregate per-component text cap, so many small annotations cannot produce an unbounded searchable metadata field.

File: src/services/componentSearchIndex.ts (hunk 1 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — author emails from published_by are no longer included in free-text metadata indexing.

File: src/services/componentSearchIndex.ts (hunk 3 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — python_dependencies is now indexed. Dependency/library names are useful search signals, while noisy large fields like python_original_code remain excluded.

File: src/services/componentSearchIndex.ts [hunk 4]
```diff
@@ -72,6 +77,66 @@ export function indexEntryToLexicalMatch(entry: IndexEntry): LexicalMatch {
   };
 }
 
+const ANNOTATION_KEYS_EXCLUDED_FROM_SEARCH = new Set([
```
Comment (maxy-shpfy) [inline src/services/componentSearchIndex.ts:80]: > 🤖 This is an AI-generated code review comment.

This PR adds annotation serialization/filtering plus a fifth searchable field to the already-central lexical search module. FTA on this file regressed from 58.52 to 62.80, cyclomatic complexity from 34 to 47, and line count from 218 to 289, crossing the str…

Suggested fix: Extract the new annotation/metadata projection into a small co-located helper (for example, `componentSearchMetadata.ts` or a local pure section with targeted exports) with the existing tests moved beside it, and keep…

**Review by maxy-shpfy — APPROVED** [review]

**Comment by Mbeaulne** [comment]
### Merge activity

* **Jun 26, 7:34 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2423).
* **Jun 26, 7:34 PM UTC**: @Mbeaulne merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2423).
